pkg/didcomm/protocol/didexchange/persistence.go

```diff
@@ -94,13 +104,81 @@ func (c *ConnectionRecorder) GetInvitation(id string) (*Invitation, error) {
 	return result, nil
 }
 
-// GetConnectionRecord return connection record
+// GetConnectionRecord return connection record based on the connection ID
 func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
-	name, err := c.store.Get(connectionID)
+	k := connectionKeyPrefix(connectionID)
+	connRecordBytes, err := c.store.Get(k)
+	if err != nil {
+		return nil, err
+	}
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// GetConnectionRecordByProtocolID return connection record via protocol ID
+func (c *ConnectionRecorder) GetConnectionRecordByProtocolID(protocolID string) (*ConnectionRecord, error) {
+	connectionIDBytes, err := c.store.Get(protocolID)
+	if err != nil {
+		return nil, err
+	}
+	// adding prefix for storing connection record
+	k := connectionKeyPrefix(string(connectionIDBytes))
+
+	connRecordBytes, err := c.store.Get(k)
 	if err != nil {
 		return nil, err
 	}
-	return &ConnectionRecord{State: string(name)}, nil
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// saveConnectionRecord saves the connection record against the connection id  in the store
+func (c *ConnectionRecorder) saveConnectionRecord(record *ConnectionRecord) error {
+	k := connectionKeyPrefix(record.ConnectionID)
+	bytes, err := json.Marshal(record)
+	if err != nil {
+		return fmt.Errorf("save connection record: %s", err)
+	}
+	return c.store.Put(k, bytes)
+}
+
+// saveNewConnectionRecord saves newly created connection record against the connection id in the store
+// and it creates mapping from protocolID to connection ID
+func (c *ConnectionRecorder) saveNewConnectionRecord(record *ConnectionRecord) error {
+	if record.ThreadID == "" || record.ConnectionID == "" || record.Namespace == "" {
+		return fmt.Errorf("input parameters thid : %s and connectionId : %s namespace : %s cannot be empty",
+			record.ThreadID, record.ConnectionID, record.Namespace)
+	}
+	err := c.saveConnectionRecord(record)
+	if err != nil {
+		return err
+	}
+	return c.saveProtocolID(record.ConnectionID, record.ThreadID, record.Namespace)
+}
+func (c *ConnectionRecorder) saveProtocolID(connectionID, thid, namespace string) error {
```

Comment:
Can we arrange the order of args here (key, val) ? connectionID is a value here (first arg), this is bit less readable.

Reply:
shuffled